File: python/Knucklebones.py
import os

from typing import List
from random import randint
from collections import Counter


class Knucklebones():
    def __init__(self) -> None:
        self.player_one_board: List[List[int]] = [[0,0,0],[0,0,0],[0,0,0]]
        self.player_two_board: List[List[int]] = [[0,0,0],[0,0,0],[0,0,0]]

        self.player_one_score: int = 0
        self.player_two_score: int = 0

        self.game_over: bool = False
        self.starting_player: int = randint(0, 1)
        self.current_player: int = self.starting_player
    
    def start(self) -> None:
        print("starting game...")
        print(f"player {self.starting_player + 1} starts.")

        while not self.game_over:
            if self.is_board_full():
                # TODO: fix this
                # winner is not always the player with the highest score
                # because one player may have a board full of low numbers
                winner = 1 if self.player_one_score > self.player_two_score else 2
                print(f"The winner is: Player {winner}")
                self.game_over = True
    
            self.display();
            roll = self.get_roll()
            
            while True:
                if self.is_board_full():
                    self.game_over = True
                    break

                column = input(f"\nplayer {self.current_player + 1}: where do you want to put your {roll}? ")
                if column.isdigit() and int(column) in [1,2,3]:
                    column_index = int(column) - 1
            
                    if self.current_player == 0:
                        if 0 in self.player_one_board[column_index]:
                            free_row = self.player_one_board[column_index].index(0) 
                            self.player_one_board[column_index][free_row] = roll
                            
                            for index, number in enumerate(self.player_two_board[column_index]):
                                if number == roll:
                                    self.player_two_board[column_index][index] = 0

                            break
                        else:
                            print("column is full")

                    if self.current_player == 1:
                        if 0 in self.player_two_board[column_index]:
                            free_row = self.player_two_board[column_index].index(0)
                            self.player_two_board[column_index][free_row] = roll
                            
                            if roll in self.player_one_board[column_index]:
                                for index, number in enumerate(self.player_one_board[column_index]):
                                    if number == roll:
                                        self.player_one_board[column_index][index] = 0
        
                            break
                        else:
                            print("column is full")
                else:
                    print("invalid column")
            
            # TODO: clear and display again to remove incorrect input prints

            self.current_player = 1 if self.current_player == 0 else 0
            
            self.clear()
            self.calculate_score()

        print("game over")
        
    def calculate_score(self) -> None:
        b1 = self.player_one_board
        b2 = self.player_two_board
        
        p1_score = 0
        p2_score = 0
        
        for column in b1:
            for number, count in Counter(column).items():
                if number > 0:
                    p1_score += number * count**2

        for column in b2:
            for number, count in Counter(column).items():
                if number > 0:
                    p2_score += number * count**2
            
        self.player_one_score = p1_score
        self.player_two_score = p2_score

    def is_board_full(self) -> bool:
        b1 = self.player_one_board
        b2 = self.player_two_board

        return not 0 in (b1[0] + b1[1] + b1[2] + b2[0] + b2[1] + b2[2])
        
    def get_roll(self) -> int:
        return randint(1, 6)
    
    @staticmethod
    def clear() -> None:
        os.system("cls" if os.name == "nt" else "clear")

    def display(self) -> None:
        b1 = self.player_one_board
        b2 = self.player_two_board
        
        print()

        print(f"Player One: {self.player_one_score}")
        print(f"| {b1[0][0]} | {b1[1][0]} | {b1[2][0]} |")
        print(f"| {b1[0][1]} | {b1[1][1]} | {b1[2][1]} |")
        print(f"| {b1[0][2]} | {b1[1][2]} | {b1[2][2]} |")
        
        print()
        
        print(f"Player Two: {self.player_two_score}")
        print(f"| {b2[0][0]} | {b2[1][0]} | {b2[2][0]} |")
        print(f"| {b2[0][1]} | {b2[1][1]} | {b2[2][1]} |")
        print(f"| {b2[0][2]} | {b2[1][2]} | {b2[2][2]} |")

File: rust/src/knucklebones.rs
use std::io::{stdin, stdout, Write};

use rand::Rng;

pub struct Knucklebones {
    player_one_board: [[i32; 3]; 3],
    player_two_board: [[i32; 3]; 3],

    player_one_score: i32,
    player_two_score: i32,

    game_over: bool,
    starting_player: i32,
    current_player: i32,
}

impl Knucklebones {
    // TODO: make static
    fn get_roll(&self) -> i32 {
        rand::thread_rng().gen_range(1..7)
    }

    // TODO: make static
    fn clear(&self) {
        print!("{}[2J", 27 as char);
        stdout().flush().unwrap();
    }

    pub fn new() -> Knucklebones {
        let starting_player = rand::thread_rng().gen_range(0..2);

        return Knucklebones {
            player_one_board: [[0; 3]; 3],
            player_two_board: [[0; 3]; 3],

            player_one_score: 0,
            player_two_score: 0,

            game_over: false,
            starting_player,
            current_player: starting_player,
        };
    }

    pub fn start(&mut self) {
        println!("starting game...");
        println!("player {} starts.", self.starting_player + 1);

        while !self.game_over {
            if self.is_board_full() {
                // TODO: fix this winner is not always the player with the highest
                //  score because one player may have a board full of low numbers
                let winner = if self.player_one_score > self.player_two_score { 1 } else { 2 };
                println!("The winner is: Player {}", winner);
                self.game_over = true;
            }

            self.display();
            let roll = self.get_roll();

            loop {
                if self.is_board_full() {
                    self.game_over = true;
                    break;
                }

                // TODO: dont clear the invalid input message
                // self.clear();
                // self.display();

                print!("\nplayer {}: where do you want to put your {}? ", self.current_player + 1, roll);
                let mut column = String::new();
                stdout().flush().unwrap();
                stdin().read_line(&mut column).unwrap();

                let column_index = match column.trim().parse::<usize>() {
                    Ok(num) => {
                        match num {
                            1..=3 => num - 1,
                            _ => {
                                println!("invalid column");
                                continue;
                            }
                        }
                    }
                    Err(_) => {
                        println!("invalid column");
                        continue;
                    }
                };

                println!("column_index: {}", column_index);

                if self.current_player == 0 {
                    if self.player_one_board[0][column_index] != 0 {
                        println!("column is full");
                        continue;
                    }

                    let free_row = self.player_one_board[column_index]
                        .iter()
                        .position(|&x| x == 0)
                        .unwrap();
                    self.player_one_board[column_index][free_row] = roll;

                    break;
                }

                self.current_player = if self.current_player == 0 { 1 } else { 0 };

                if self.current_player == 1 {
                    if self.player_two_board[0][column_index] != 0 {
                        println!("column is full");
                        continue;
                    }

                    let free_row = self.player_two_board[column_index]
                        .iter()
                        .position(|&x| x == 0)
                        .unwrap();
                    self.player_two_board[column_index][free_row] = roll;

                    break;
                }
            }

            self.current_player = if self.current_player == 0 { 1 } else { 0 };

            // self.clear();
            self.calculate_score();
        }
        println!("game over!");
    }

    fn is_board_full(&self) -> bool {
        for row in 0..3 {
            for col in 0..3 {
                if self.player_one_board[col][row] == 0 {
                    return false;
                }
            }
        }
        return true;
    }

    fn calculate_score(&mut self) {
        let mut player_one_score = 0;
        let mut player_two_score = 0;

        for row in 0..3 {
            for col in 0..3 {
                if self.player_one_board[row][col] == 1 {
                    player_one_score += 1;
                }
                if self.player_two_board[row][col] == 1 {
                    player_two_score += 1;
                }
            }
        }

        self.player_one_score = player_one_score;
        self.player_two_score = player_two_score;
    }

    fn display(&self) {
        println!("\nPlayer One: {}", self.player_one_score);
        for row in 0..3 {
            for col in 0..3 {
                print!("| {} ", self.player_one_board[col][row]);
            }
            print!("|\n");
        }

        println!("\nPlayer Two: {}", self.player_two_score);
        for row in 0..3 {
            for col in 0..3 {
                print!("| {} ", self.player_two_board[col][row]);
            }
            print!("|\n");
        }
    }
}